seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/sink/client/executor/InsertOrUpdateBatchStatementExecutor.java

```diff
@@ -0,0 +1,138 @@
+ */
+
+package org.apache.seatunnel.connectors.seatunnel.clickhouse.sink.client.executor;
+
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.function.Function;
+
+@RequiredArgsConstructor
+public class InsertOrUpdateBatchStatementExecutor implements JdbcBatchStatementExecutor {
+    private final StatementFactory existStmtFactory;
+    @NonNull
+    private final StatementFactory insertStmtFactory;
+    @NonNull
+    private final StatementFactory updateStmtFactory;
+    private final Function<SeaTunnelRow, SeaTunnelRow> keyExtractor;
+    private final JdbcRowConverter keyRowConverter;
+    @NonNull
+    private final JdbcRowConverter valueRowConverter;
+    private transient PreparedStatement existStatement;
+    private transient PreparedStatement insertStatement;
+    private transient PreparedStatement updateStatement;
+    private transient Boolean preChangeFlag;
+    private transient boolean submitted;
+
+    public InsertOrUpdateBatchStatementExecutor(StatementFactory insertStmtFactory,
+                                                StatementFactory updateStmtFactory,
+                                                JdbcRowConverter rowConverter) {
+        this(null, insertStmtFactory, updateStmtFactory,
+            null, null, rowConverter);
+    }
+
+    @Override
+    public void prepareStatements(Connection connection) throws SQLException {
+        if (upsertMode()) {
+            existStatement = existStmtFactory.createStatement(connection);
+        }
+        insertStatement = insertStmtFactory.createStatement(connection);
+        updateStatement = updateStmtFactory.createStatement(connection);
+    }
+
+    private boolean upsertMode() {
+        return existStmtFactory != null;
+    }
+
+    private boolean hasInsert(SeaTunnelRow record) throws SQLException {
+        if (upsertMode()) {
+            return exist(keyExtractor.apply(record));
+        }
+        switch (record.getRowKind()) {
+            case INSERT:
+                return true;
+            case UPDATE_AFTER:
+                return false;
+            default:
+                // todo
+                throw new UnsupportedOperationException();
+        }
+    }
+
+    @Override
+    public void addToBatch(SeaTunnelRow record) throws SQLException {
+        boolean currentChangeFlag = hasInsert(record);
+        if (currentChangeFlag) {
+            if (preChangeFlag != null && !preChangeFlag) {
+                insertStatement.executeBatch();
+                insertStatement.clearBatch();
+            }
+            valueRowConverter.toExternal(record, updateStatement);
+            updateStatement.addBatch();
+        } else {
+            if (preChangeFlag != null && preChangeFlag) {
```

Comment:
If the previous record is an update and this record is an insert, it seems unnecessary to execute the update statement.